libs/libarchfpga/src/physical_types_util.h

```diff
@@ -0,0 +1,132 @@
+
+int find_pin(t_physical_tile_type_ptr type, std::string port_name, int pin_index_in_port);
+
+//Returns the maximum number of pins within a logical block
+int get_max_num_pins(t_logical_block_type_ptr logical_block);
+
+//Verifies whether a given logical block is compatible with a given physical tile
+bool is_tile_compatible(t_physical_tile_type_ptr physical_tile, t_logical_block_type_ptr logical_block);
+
+//Verifies whether a logical block and a relative placement location is compatible with a given physical tile
+bool is_sub_tile_compatible(t_physical_tile_type_ptr physical_tile, t_logical_block_type_ptr logical_block, int sub_tile_loc);
+
+//Returns the first physical tile type that matches the logical block
+//
+//The order of the physical tiles suitable for the input logical block follows the order
+//with which the logical blocks appear in the architecture XML definition
+t_physical_tile_type_ptr pick_physical_type(t_logical_block_type_ptr logical_block);
+
+//Returns the first logical block type that matches the physical tile
+//
+//The order of the logical blocks suitable for the input physical tile follows the order
+//with which the physical tiles appear in the architecture XML definition
+t_logical_block_type_ptr pick_logical_type(t_physical_tile_type_ptr physical_tile);
+
+//Returns the sub tile index (within 'physical_tile') corresponding to the
+//'logical block'.
+//
+//This function will return the index for the first sub_tile that can accommodate
+//the logical block.
+//
+//It is typically called before/during placement,
+//when picking a sub-tile to fit a logical block
+int get_logical_block_physical_sub_tile_index(t_physical_tile_type_ptr physical_tile,
+                                              t_logical_block_type_ptr logical_block);
+
+//Returns the physical pin index (within 'physical_tile') corresponding to the
+//logical index ('pin' of the first instance of 'logical_block' within the physcial tile.
+//
+//This function is called before/during placement, when a sub tile index was not yet assigned.
+//
+//Throws an exception if the corresponding physical pin can't be found.
+int get_physical_pin(t_physical_tile_type_ptr physical_tile,
+                     t_logical_block_type_ptr logical_block,
+                     int pin);
+
+//Returns the physical pin index (within 'physical_tile') corresponding to the
+//logical index ('pin' of the first instance of 'logical_block' within the physcial tile.
+//This function considers if a given offset is in the range of sub tile capacity
+//
+//  (First pin index at current sub-tile)                                     (The wanted pin index)
+//
+//  |                                                               |<----- pin ------->|
+//  v                                                                                   v
+//
+//  |<----- capacity.low ----->|<----- capacity.low + 1 ----->| ... |<----- sub_tile_capacity ---->|
```

Comment:
I like the example here but still don't fully understand it.  Can you make it more concrete or add a concrete example?

Reply:
Done